### PR TITLE
Toolkit feature

### PR DESCRIPTION
**Pull request by luciferlinx101**
- Toolkit  
- Toolkit Configuration
- Google Calendar

### CODE REVIEW

File: superagi/helper/google_calendar_creds.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        if os.path.exists(file_path):
+            engine = connect_db()
+            Session = sessionmaker(bind=engine)
+            session = Session()
+            with open(file_path,'rb') as file:
+                creds = pickle.load(file)
+            if isinstance(creds, str):
+                creds = json.loads(creds)
+            expire_time = datetime.strptime(creds["expiry"], "%Y-%m-%dT%H:%M:%S.%fZ")
+            toolkit_id = 14
```
Comment (TransformerOptimus) [inline superagi/helper/google_calendar_creds.py:34]: what is 14?

File: superagi/helper/google_calendar_creds.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        file_name = "credential_token.pickle"
+        root_dir = get_config('RESOURCES_OUTPUT_ROOT_DIR')
+        file_path = file_name
+        if root_dir is not None:
+            root_dir = root_dir if root_dir.startswith("/") else os.getcwd() + "/" + root_dir
+            root_dir = root_dir if root_dir.endswith("/") else root_dir + "/"
+            file_path = root_dir + file_name
+        else:
+            file_path = os.getcwd() + "/" + file_name
+        if os.path.exists(file_path):
```
Comment (TransformerOptimus) [inline superagi/helper/google_calendar_creds.py:25]: Too large method. break into multiple methods

File: superagi/controllers/agent.py [hunk 3]
```diff
@@ -157,9 +158,22 @@ def create_agent_with_config(agent_with_config: AgentWithConfig,
         if tool is None:
             # Tool does not exist, throw 404 or handle as desired
             raise HTTPException(status_code=404, detail=f"Tool with ID {tool_id} does not exist. 404 Not Found.")
+
+    agent_tool_kit_tools = []
+    for tool_kit_id in agent_with_config.tool_kits:
```
Comment (TransformerOptimus) [inline superagi/controllers/agent.py:163]: move this to toolkit model

File: superagi/controllers/tool.py [hunk 4]
```diff
@@ -66,6 +68,18 @@ def get_tool(
     return db_tool
 
 
+@router.get("/get")
+def get_tools(Authorize: AuthJWT = Depends(check_auth),
+              organisation: Organisation = Depends(get_user_organisation)):
+    """Get all tools"""
+    tool_kits = db.session.query(ToolKit).filter(ToolKit.organisation_id == organisation.id)
+    tools = []
+    for tool_kit in tool_kits:
```
Comment (TransformerOptimus) [inline superagi/controllers/tool.py:77]: rename tool_kit to toolkit.

File: superagi/controllers/tool.py [hunk 5]
```diff
@@ -66,6 +68,18 @@ def get_tool(
     return db_tool
 
 
+@router.get("/get")
+def get_tools(Authorize: AuthJWT = Depends(check_auth),
```
Comment (TransformerOptimus) [inline superagi/controllers/tool.py:72]: get_user_organisation does auth as well. why do u need check_auth again?

File: superagi/helper/calendar_date.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
+from datetime import datetime, timedelta, timezone
+import pytz
+from dateutil import tz
+
+class CalendarDate():
+    def get_date_utc(self,start_date,end_date,start_time,end_time,service):
```
Comment (TransformerOptimus) [inline superagi/helper/calendar_date.py:6]: Can we refactor get_date_utc & create_event_dates better way

File: superagi/helper/tool_helper.py [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+
+
+def init_tool_kits(code_link, existing_toolkits, folder_path, organisation, session):
+    tool_name_to_tool_kit = {}
+    new_toolkits = []
+    # Iterate over all subfolders
+    for folder_name in os.listdir(folder_path):
+        folder_dir = os.path.join(folder_path, folder_name)
+
+        if os.path.isdir(folder_dir):
```
Comment (TransformerOptimus) [inline superagi/helper/tool_helper.py:145]: break this into multiple methods.

File: superagi/helper/tool_helper.py [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+    module = importlib.util.module_from_spec(spec)
+    spec.loader.exec_module(module)
+
+    return module
+
+
+def init_tools(folder_path, session, tool_name_to_tool_kit):
+    print(tool_name_to_tool_kit)
+    # Iterate over all subfolders
+    for folder_name in os.listdir(folder_path):
```
Comment (TransformerOptimus) [inline superagi/helper/tool_helper.py:114]: break this into multiple methods.

File: superagi/helper/tool_helper.py [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+def load_module_from_file(file_path):
+    spec = importlib.util.spec_from_file_location("module_name", file_path)
+    module = importlib.util.module_from_spec(spec)
+    spec.loader.exec_module(module)
+
+    return module
+
+
+def init_tools(folder_path, session, tool_name_to_tool_kit):
+    print(tool_name_to_tool_kit)
```
Comment (TransformerOptimus) [inline superagi/helper/tool_helper.py:112]: use logger instead

File: superagi/helper/tool_helper.py [hunk 10]
```diff
@@ -0,0 +1,250 @@
...
+
+            if not target_name:
+                continue
+
+            if member.endswith('/'):
+                os.makedirs(target_path, exist_ok=True)
+            else:
+                with open(target_path, 'wb') as outfile, z.open(member) as infile:
+                    outfile.write(infile.read())
+    print("Donwload Success!")
```
Comment (TransformerOptimus) [inline superagi/helper/tool_helper.py:66]: use logger

File: superagi/tools/email/email_tool_kit.py [hunk 11]
```diff
@@ -0,0 +1,18 @@
+from abc import ABC
+from superagi.tools.base_tool import BaseToolKit, BaseTool
+from typing import Type, List
+from read_email import ReadEmailTool
+from send_email import SendEmailTool
+from send_email_attachment import SendEmailAttachmentTool
+
+
+class EmailToolKit(BaseToolKit, ABC):
```
Comment (TransformerOptimus) [inline superagi/tools/email/email_tool_kit.py:9]: 2 tool kits for email. Remove one file

File: superagi/helper/tool_helper.py (hunk 9 above)
Comment (luciferlinx101) [reply superagi/helper/tool_helper.py]: changed

File: superagi/helper/tool_helper.py (hunk 10 above)
Comment (luciferlinx101) [reply superagi/helper/tool_helper.py]: changed

File: superagi/tools/email/email_tool_kit.py (hunk 11 above)
Comment (luciferlinx101) [reply superagi/tools/email/email_tool_kit.py:9]: removed

**Pull request closed without merging**